core/partialevaluators.go

```diff
@@ -91,63 +69,55 @@ func (policyEvaluators PartialResultsEvaluators) AddFromConfig(ctx context.Conte
 	return nil
 }
 
-func (partialEvaluators PartialResultsEvaluators) GetEvaluatorFromPolicy(ctx context.Context, policy string, input []byte, options *OPAEvaluatorOptions) (*OPAEvaluator, error) {
+func (partialEvaluators PartialResultsEvaluators) GetEvaluatorFromPolicy(ctx context.Context, policy string, options *OPAEvaluatorOptions) (*OPAEvaluator, error) {
 	if options == nil {
 		options = &OPAEvaluatorOptions{}
 	}
 
 	if eval, ok := partialEvaluators[policy]; ok {
-		inputTerm, err := ast.ParseTerm(string(input))
-		if err != nil {
-			return nil, fmt.Errorf("%w: %v", ErrFailedInputParse, err)
-		}
-
-		evaluator := eval.PartialEvaluator.Rego(
-			rego.ParsedInput(inputTerm.Value),
-			rego.EnablePrintStatements(options.EnablePrintStatements),
-			rego.PrintHook(NewPrintHook(os.Stdout, policy)),
-		)
-
 		return &OPAEvaluator{
-			PolicyName:      policy,
-			PolicyEvaluator: evaluator,
+			PolicyName: policy,
 
-			context:     ctx,
-			mongoClient: options.MongoClient,
+			evaluator:     eval,
+			context:       ctx,
+			mongoClient:   options.MongoClient,
+			generateQuery: eval.preparedPartialQuery != nil,
 		}, nil
 	}
 	return nil, fmt.Errorf("%w: %s", ErrEvaluatorNotFound, policy)
 }
 
-func newPartialResultEvaluator(ctx context.Context, policy string, opaModuleConfig *OPAModuleConfig, evaluatorOptions *OPAEvaluatorOptions) (*rego.PartialResult, error) {
-	if evaluatorOptions == nil {
-		evaluatorOptions = &OPAEvaluatorOptions{}
-	}
-	if opaModuleConfig == nil {
-		return nil, fmt.Errorf("OPAModuleConfig must not be nil")
-	}
+func createPartialEvaluator(logger logging.Logger, policy string, opaModuleConfig *OPAModuleConfig, options *OPAEvaluatorOptions, isPartial bool) (*PartialEvaluator, error) {
+	logger.WithField("policyName", policy).Info("precomputing rego policy")
 
-	sanitizedPolicy := strings.Replace(policy, ".", "_", -1)
-	queryString := fmt.Sprintf("data.policies.%s", sanitizedPolicy)
-
-	options := []func(*rego.Rego){
-		rego.Query(queryString),
-		rego.Module(opaModuleConfig.Name, opaModuleConfig.Content),
-		rego.Unknowns(Unknowns),
-		rego.EnablePrintStatements(evaluatorOptions.EnablePrintStatements),
-		rego.PrintHook(NewPrintHook(os.Stdout, policy)),
-		rego.Capabilities(ast.CapabilitiesForThisVersion()),
-		custom_builtins.GetHeaderFunction,
-	}
-	if evaluatorOptions.MongoClient != nil {
-		ctx = custom_builtins.WithMongoClient(ctx, evaluatorOptions.MongoClient)
-		options = append(options, custom_builtins.MongoFindOne, custom_builtins.MongoFindMany)
+	preparedPartialEvaluator := &PartialEvaluator{}
+
+	policyEvaluatorTime := time.Now()
+
+	regoInstance, err := newRegoInstanceBuilder(policy, opaModuleConfig, options)
+	if err != nil {
+		return nil, err
 	}
-	if evaluatorOptions.Logger != nil {
-		ctx = logging.WithContext(ctx, evaluatorOptions.Logger)
+	if isPartial {
+		preparedPartialQuery, err := regoInstance.PrepareForPartial(context.TODO())
```

Comment:
why `context.TODO()` ? Is this an oversight, or on purpose because this function does not receive a ctx anymore?

Reply:
Too much cleanup 😄  Added the context.